### PR TITLE
[android][barcode-scanner] Return raw barcode value from scan results

### LINKED ISSUE #23977: [expo-barcode-scanner][android] Wi-Fi QR code does not return correct data
### Minimal reproducible example

https://snack.expo.dev/@robrechtme/test-barcode-scanner

### Summary

When scanning a Wi-Fi QR code on Android using Expo SDK 49, the `data` param contains `My Network my password` instead of the expected `WIFI:S:My Network;H:true;T:WPA2;P:my password;;`

This works as expected on iOS and on SDK 48 (both platforms).

<details>

<summary>Minimal example</summary>

```jsx
import { useState, useEffect } from "react";
import { Text, View, StyleSheet, Button } from "react-native";
import { BarCodeScanner } from "expo-barcode-scanner";

export default function App() {
  const [hasPermission, setHasPermission] = useState(null);
  const [scanned, setScanned] = useState(false);

  useEffect(() => {
    const getBarCodeScannerPermissions = async () => {
      const { status } = await BarCodeScanner.requestPermissionsAsync();
      setHasPermission(status === "granted");
    };

    getBarCodeScannerPermissions();
  }, []);

  const handleBarCodeScanned = ({ type, data }) => {
    setScanned(true);
    alert(`Bar code with type ${type} and data ${data} has been scanned!`);
  };

  if (hasPermission === null) {
    return <Text>Requesting for camera permission</Text>;
  }
  if (hasPermission === false) {
    return <Text>No access to camera</Text>;
  }

  return (
    <View style={styles.container}>
      <BarCodeScanner
        onBarCodeScanned={scanned ? undefined : handleBarCodeScanned}
        style={StyleSheet.absoluteFillObject}
      />
      {scanned && (
        <Button title={"Tap to Scan Again"} onPress={() => setScanned(false)} />
      )}
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: "column",
    justifyContent: "center",
  },
});
```

</details>

<details>

<summary>Example QR code</summary>

image

</details>

### Environment

```
  expo-env-info 1.0.5 environment info:
    System:
      OS: macOS 13.5
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 18.16.1 - ~/.asdf/installs/nodejs/18.16.1/bin/node
      Yarn: 1.22.19 - ~/.asdf/shims/yarn
      npm: 9.5.1 - ~/.asdf/plugins/nodejs/shims/npm
      Watchman: 2023.06.12.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.11.3 - /Users/robrechtme/.asdf/shims/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 22.4, iOS 16.4, macOS 13.3, tvOS 16.4, watchOS 9.4
    IDEs:
      Android Studio: 2022.3 AI-223.8836.35.2231.10406996
      Xcode: 14.3.1/14E300c - /usr/bin/xcodebuild
    npmPackages:
      expo: ~49.0.0 => 49.0.7 
      react: 18.2.0 => 18.2.0 
      react-native: 0.72.3 => 0.72.3 
    Expo Workflow: managed
```

### PR DESCRIPTION
**Pull request by ajacquierbret**
# Why

Closes #23977
When scanning some barcode types (such as Wi-Fi QR codes), MLKit automatically parses the output and store the result in `barcode.displayValue`, sometimes in a very inconvenient way (see: https://developers.google.com/android/reference/com/google/mlkit/vision/barcode/common/Barcode#public-string-getdisplayvalue). Because the `expo-barcode-scanner` package only returns the parsed `displayValue` from barcode scan results, developers can't handle the parsing themselves.

# How

I added a `raw` field returning the raw encoded value from barcode scan results through the `onBarCodeScanned` callback.

# Test Plan

I haven't been able to run jest test suites locally, but because the behavior of existing fields remain untouched, there should be no breaking changes.

### CODE REVIEW

File: packages/expo-barcode-scanner/src/BarCodeScanner.tsx [hunk 1]
```diff
@@ -62,9 +62,15 @@ export type BarCodeScannerResult = {
...
-   * The information encoded in the bar code.
+   * The parsed information encoded in the bar code.
    */
   data: string;
+  /**
+   * The raw information encoded in the bar code.
+   * May be different from `data` depending on the barcode type.
+   * Will always be undefined on iOS.
+   */
+  raw: string;
```
Comment (lukmccall) [inline packages/expo-barcode-scanner/src/BarCodeScanner.tsx:73]:
```suggestion
  raw?: string;
```

**Review by lukmccall — APPROVED** [review]
Looking good ✅

File: packages/expo-barcode-scanner/src/BarCodeScanner.tsx [hunk 2]
```diff
@@ -62,9 +62,15 @@ export type BarCodeScannerResult = {
...
   type: string;
   /**
-   * The information encoded in the bar code.
+   * The parsed information encoded in the bar code.
    */
   data: string;
+  /**
+   * The raw information encoded in the bar code.
+   * May be different from `data` depending on the barcode type.
+   * Will always be undefined on iOS.
```
Comment (Kudo) [inline packages/expo-barcode-scanner/src/BarCodeScanner.tsx:71]:
```suggestion
   @platform android
```

it's quite advanced feature, i would like to add `@hidden` to hide this from the [doc](https://docs.expo.dev/versions/latest/sdk/bar-code-scanner/). how do people think about this?

Comment (ajacquierbret) [reply packages/expo-barcode-scanner/src/BarCodeScanner.tsx]: @Kudo Is the `@hidden` flag meant to completely hide the field reference and documentation?

File: packages/expo-barcode-scanner/CHANGELOG.md [hunk 3]
```diff
@@ -10,6 +10,8 @@
 
 ### 💡 Others
 
+- `BarCodeScannerResult` now returns an additional `raw` field corresponding to the barcode value as it was encoded in the barcode without parsing. Will always be undefined on iOS.
```
Comment (tsapeta) [inline packages/expo-barcode-scanner/CHANGELOG.md:13]: I think it makes more sense to move this a few lines up to new features section 😉

File: packages/expo-barcode-scanner/src/BarCodeScanner.tsx (hunk 2 above)
Comment (Kudo) [reply packages/expo-barcode-scanner/src/BarCodeScanner.tsx]: it should expose to typescript with autocomplete and show the property comment if mouseover the property. just not into the doc.

**Comment by ajacquierbret** [comment]
@Kudo @tsapeta I committed your suggestions, the PR should now be ready for a merge

**Comment by ajacquierbret** [comment]
> Nice! Imo it would make sense to support this on iOS as well, returning the same value as `data`.

From my understanding iOS doesn't automatically parses the scan results, so it shouldn't be needed.

**Comment by lukmccall** [comment]
The CI is failing on `check-packages`. Can you rebuild the js files?

**Comment by ajacquierbret** [comment]
@Kudo @tsapeta @lukmccall @alanjhughes JS files have been rebuilt, changelogs have been updated. The PR should be ready.

**Comment by alanjhughes** [comment]
@ajacquierbret - As this library will soon be deprecated, would you like to add this to the built in scanner in `expo-camera/next`? 
Also, you'll need to run spotless on android `./gradlew :expo-barcode-scanner:spotlessApply` and run `git submodule update`

**Comment by ajacquierbret** [comment]
@alanjhughes I'll do that. Thanks for your reply!

**Comment by ajacquierbret** [comment]
@alanjhughes I tried for hours to run spotless but didn't manage to do it, it seems that `apps/expo-go/android/settings.gradle:25` references a missing file, and I faced various issues trying to run this command from other apps. Git submodules (i.e. `react-native`) have been properly updated through `yarn setup:native`.
Could you have a look?

Apart from that, all relevant changes have been applied to `expo-camera/next` so it should now be only a matter of CI checks to pass before merging the PR.

**Comment by alanjhughes** [comment]
Hi @ajacquierbret - You shouldn't need to do anything with expo go. cd into the android directory of `bare-expo` and that should work 🤔

**Comment by ajacquierbret** [comment]
Hi @alanjhughes

I managed to run spotless from `bare-expo`, it previously failed because I didn't have Android NDK 25.1.8937393 installed locally, as I thought that running `setup:native` would take care of it as mentioned in the `CONTRIBUTING.md` guide.

I tried to run spotless from `expo-go` because that's the step where the CI fails.
But it keeps failing because `apps/expo-go/android/settings.gradle:25` seems to reference a missing file, and as I said earlier, I don't know why.

**Comment by alanjhughes** [comment]
Hi @ajacquierbret the issue is with the submodule. Have you run `git submodule update`? It should not be different to what is on main.

**Comment by ajacquierbret** [comment]
Hi @alanjhughes - It was a submodule issue.

Please note that `CONTRIBUTING.md` recommends to run the `setup:native` command that executes `scripts/download-dependencies.sh` in order to update submodules, but then `react-native` points to an old commit, not the `exp-latest` branch. I had to checkout to `exp-latest` manually from the submodule. It looks like a script issue.

Anyway, it should be fine now.

**Review by alanjhughes — APPROVED** [review]

**Comment by alanjhughes** [comment]
Thank you for your contribution @ajacquierbret!

**Pull request merged**